cli/azd/design-notes/environments.md

```diff
@@ -0,0 +1,271 @@
+param principalId string = ''
+```
+
+And a corresponding `.parameters.json` file under the `infra` folder to tie these values to what was configured in the environment:
+
+```json
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "name": {
+      "value": "${AZURE_ENV_NAME}"
+    },
+    "location": {
+      "value": "${AZURE_LOCATION}"
+    },
+    "principalId": {
+      "value": "${AZURE_PRINCIPAL_ID}"
+    }
+  }
+}
+```
+
+During `azd provision` the following steps happen:
+
+1. The `main.parameters.json` file is loaded, and evaluated in the current environment (so all the `${}` expressions are replaced with actual values).
+2. The `main.bicep` file is loaded and all the deployment parameters are collected.
+3. If any required parameters are missing, `azd` prompts for their values.
+4. A new `main.parameters.json` file is written in the environment directory (`.azure/<env-name>/`) which contains the final configured values for the parameters.
+5. An ARM Deployment is created to deploy the bicep template with the `main.parameters.json` file from step 4. The name of this deployment matches the environment name and the target subscription is pulled from the `AZURE_SUBSCRIPTION_ID` environment variable.
+6. As described earlier, when the deployment completes, all outputs are merged in to the `.env` file for the environment.
+
+By convention, our templates also tag resources with `azd-env-name` and there has been discussion about `azd` using that tag when running queries to discover resources. We've done something similar with `azd-service-name` and that pattern has worked out well for us and plays nicely in the infrastructure as code world.
+
+This system allows you to create a multiple environments, each with a different name, and provision a unique copy of the infrastructure for the application.
+
+## Current Challenges
+
+The major issue we face with the current design is that we've introduced a new concept, an environment, and it is on the critical path of every `azd` invocation. We need to ensure this concept is simple to understand and worthwhile and clearly articulate how it relates to the the rest of the system (and ideally existing Azure concepts). It's not always clear how settings in the environment impact infrastructure or the rest of `azd` and what's based on convention and what is required by the tooling. There are also "tiers" of configuration values now - `AZURE_ENV_NAME`, `AZURE_LOCATION`, `AZURE_PRINCIPAL_ID` and `AZURE_SUBSCRIPTION_ID` are very different from other values. We also have a tension on how to describe what an environment name does. Our templates today use it to influence the names of some resources (usually the root resource group that is created) but that's simply a convention we've adopted and so we are wary of adding that as "help text" for an environment name.
```

Comment:
Let's fully flesh out what the current challenges are.  Is the challenge not being able to clearly articulate what an environment is?

An environment is metadata that describes an Azure application.

We could change the name to 'project' or 'solution' if that helps.

Reply:
Well, there is no such thing as "Azure application" today, only applications that can be deployed to Azure. 

Our environment holds metadata about an Azure deployment of an application. It contains both "input" data, that is, data that is necessary for AZD CLI to deploy the application, as well as "output" data, that is, data that is a result of a deployment. This "input vs output" difference, and the fact that the user can add extra data to the environment, data that is neither necessary for deployment, nor a result of a deployment, plays a role in making the environment a bit confusing.

"Project" and "solution" have a very specific meaning in the dev tooling world, and calling environments either "project" or "solution" will increase the confusion quite a bit, let's not do that please.

Reply:
I'm still not convinced that `azd` is introducing something novel -- environment as a concept is an industry-wide term used for developers that work on services.

The definition of an environment for me:

A collection of resources that form an instance of the application service. This can be used to isolate non-production vs production instances, or to create regional instances when necessary.

And I don't think `azd` is doing something different ahn that.

Reply:
@weikanglim respectfully disagree. In your definition ("a collection of resources that form an instance of the application service") I assume you mean "a container". If that is the case, you are describing a cloud-side, or server-side concept: a resource group, datacenter, Kubernetes cluster etc. 

This is different from azd environment, which (today) is a client-side concept with no cloud-side counterpart. It is just a named set of metadata related to application deployment.

Reply:
@karolz-ms 

> In your definition ("a collection of resources that form an instance of the application service") I assume you mean "a container". If that is the case, you are describing a cloud-side, or server-side concept: a resource group, datacenter, Kubernetes cluster etc.

I think we mean the same, but just to clarify -- "A collection of resources" here is more abstract than server-side groupings. It simply states that we have a set of resources that when formed together, becomes the application that is being delivered. It can be multiple resource groups, multiple datacenters or clusters, or perhaps just a single VM.

> This is different from azd environment, which (today) is a client-side concept with no cloud-side counterpart. It is just a named set of metadata related to application deployment.

Yes, I was not being technically correct here. I was focused on the actual usage of the metadata because that's what customers use `azd` environments to do, rather than describing what an `azd environment` is in terms of what it holds. Let me attempt at redefining it:

An environment describes a collection of resources that form the application and can be deployed. Typical environments include: development, test, staging or production.

I think this is very similar to how Azure DevOps define their environments, which is also metadata that ADO stores on their service that facilitates multi-stage deployments: https://docs.microsoft.com/en-us/azure/devops/pipelines/process/environments?view=azure-devops

Reply:
I guess it is useful to differentiate between what an environment IS vs what it is USED FOR. From the latter perspective it is used to provision/deploy separate instances of an application, possibly on a different scale, different access permissions, location etc. etc. (although today `azd` does not help you much with all these aspects). I think this is what you are describing and I agree we need to include this information when we talk about AZD environments.

It is also worth noting that today `azd` environment can really only manage a single application instance. This is different from AzDO environments or, say, Azure Container Applications environments.